### ISSUE TITLE
in KF0.6.2 IAP GKE install, 'iap-enabler' pod is crash-looping

### ISSUE THREAD

**Issue by amygdala** [kubeflow/kubeflow#4043]
/kind bug

using KF0.6.2 w/ IAP on GKE.

I see the iap-enabler-* pod continuously crash-looping.  Here's its logs:

```
kubectl logs iap-enabler-778889874f-4vrfd -n istio-system
Activated service account credentials for: [kf062test-admin@xxxx.iam.gserviceaccount.com]
[component_manager]
disable_update_check = true
[core]
account = kf062test-admin@xxxx.iam.gserviceaccount.com
disable_usage_reporting = true
project =xxxx
[metrics]
environment = github_docker_image

Your active configuration is: [default]
node port is 31380
fetching backends info with envoy-ingress: {"k8s-be-31380--5810244e269b1626":"HEALTHY","k8s-be-32253--5810244e269b1626":"HEALTHY"}
backend name is k8s-be-31380--5810244e269b1626
Waiting for backend id PROJECT=xxxx NAMESPACE=istio-system SERVICE=istio-ingressgateway filter=name~k8s-be-31380--5810244e269b1626
BACKEND_ID=6347129120648147983
patch JWT audience: /projects/xxxx/global/backendServices/6347129120648147983
policy.authentication.istio.io/ingress-jwt not patched
Clearing lock on service annotation
service/istio-ingressgateway not patched
Fri Aug 30 17:48:10 UTC 2019 WARN: Backend check failed, restarting container.
```

I saw this w/ both kfctl-based install and using the (staging) c2d webapp.

**Comment by jlewi** [comment]
Links to source
https://github.com/kubeflow/manifests/blob/56e2fb15db286198f7a53723cb1fbfecf3fe83fb/gcp/iap-ingress/base/deployment.yaml#L44

https://github.com/kubeflow/manifests/blob/56e2fb15db286198f7a53723cb1fbfecf3fe83fb/gcp/iap-ingress/base/config-map.yaml#L127

It looks like there may have been a problem patching the services.

If anyone encounters this issue it would be great to grab the following output as well.

```
NAMESPACE=istio-system
SERVICE=istio-ingressgateway

kubectl -n ${NAMESPACE} get policy ingress-jwt -o yaml
```

It looks like we clear the annotation on the service here
https://github.com/kubeflow/manifests/blob/56e2fb15db286198f7a53723cb1fbfecf3fe83fb/gcp/iap-ingress/base/config-map.yaml#L111

We should add the -x option so we can see what the output of the commands are to aid debugging.

**Comment by jlewi** [comment]
It looks like checkBackend is checking if the backend timeout is updated to 1 hour.
https://github.com/kubeflow/manifests/blob/56e2fb15db286198f7a53723cb1fbfecf3fe83fb/gcp/iap-ingress/base/config-map.yaml#L122

But setup.sh doesn't actually update the timeout to one hour; that happens in in update_backend.sh.
So maybe we should remove that check?

So hypothesis is that its update_backend.sh that is failing to update the timeout and that's why the iap-enabler pod is crashing.

**Comment by jlewi** [comment]
We should aim to do the following

1. Try to troubleshoot this and come up with docs for explaining how users can figure out what the root cause is and fix it
1. For 0.7 add -e  to the bash scripts so we echo commands to aid debugging
1. For 0.7 remove the check in checkBackend that verifies the timeout is set correctly from setup.sh

**Comment by mesaverde0207** [comment]
HI, When I was running Kubeflow 0.6.2 in GCP according to the [CLI deployment tutorial](https://www.kubeflow.org/docs/gke/deploy/deploy-cli/), I encountered the same problem that iap-enabler-778889874f-vxtx8 was in CrashLoopBackOff.

By the suggested commands
```
NAMESPACE=istio-system
SERVICE=istio-ingressgateway

kubectl -n ${NAMESPACE} get policy ingress-jwt -o yaml
```

I have got a yaml file:
```yaml
apiVersion: authentication.istio.io/v1alpha1
kind: Policy
metadata:
  creationTimestamp: "2019-09-11T03:37:50Z"
  generation: 1
  labels:
    kustomize.component: iap-ingress
  name: ingress-jwt
  namespace: istio-system
  resourceVersion: "8275"
  selfLink: /apis/authentication.istio.io/v1alpha1/namespaces/istio-system/policies/ingress-jwt
  uid: 87b73d46-d445-11e9-ba4d-42010a800030
spec:
  origins:
  - jwt:
      audiences:
      - /projects/26155353191/global/backendServices/4716516933859922628
      issuer: https://cloud.google.com/iap
      jwksUri: https://www.gstatic.com/iap/verify/public_key-jwk
      jwtHeaders:
      - x-goog-iap-jwt-assertion
      trigger_rules:
      - excluded_paths:
        - exact: /healthz
        - prefix: /.well-known/acme-challenge
  principalBinding: USE_ORIGIN
  targets:
  - name: istio-ingressgateway
    ports:
    - number: 80
```

I hope this will be some help for this problem, thank you.

**Comment by y4roslav** [comment]
Is there some workaround  for the issue?

**Comment by mesaverde0207** [comment]
I have no idea, but Kubeflow v0.6.2 deployment by the UI tool has come to go well.
Now Kubeflow UI is working well although iap-enabler is still in CrashLoopBackOff.

**Comment by marcoleewow** [comment]
I still get iap-enabler in CrashLoopBackOff when using deploy via UI.

**Comment by mattnworb** [comment]
We are having the same problem. I believe the problem is that https://github.com/kubeflow/manifests/blob/e97671c579e176bca9928c8cefa38c1c16432fe8/gcp/iap-ingress/base/config-map.yaml#L122 requires that the `timeoutSec` of the backend service is 3600 seconds, but there is an error when `update_backend.sh` (invoked by the backend-updater StatefulSet) [tries to update the timeout to be 3600 seconds](https://github.com/kubeflow/manifests/blob/e97671c579e176bca9928c8cefa38c1c16432fe8/gcp/iap-ingress/base/config-map.yaml#L194-L199).

From the `backend-updater-0` pod's logs:

```
Increasing backend timeout for JupyterHub
ERROR: (gcloud.compute.backend-services.update) unrecognized arguments:
  https://www.googleapis.com/compute/v1/projects/our-project/global/backendServices/k8s-be-31380--912a1fa9425e39fa
  https://www.googleapis.com/compute/v1/projects/our-project/global/backendServices/k8s-be-31380--dc141d6075541ed7
Backend updated successfully. Waiting 1 hour before updating again.
```

**Comment by jlewi** [comment]
## Work around

To work around this issue I would first verify whether the crash looping is actually affecting you

These scripts are doing two things

1. Setting the JWT on the ISTIO policy
1. Updating the timeout for the backend

So if both of these are true then you should be good to go.

To check the JWT ISTIO policy

```
kubectl -n istio-system get policy ingress-jwt -o yaml
```

* If the audience is correctly set e.g. to `/projects/${PROJECTNUMBER}/global/backendServices/${BACKENDSERVICE}`

* Then your policy is correctly set
* If not then you could manually update using kubectl

To check the backend timeout

```
gcloud --project=${PROJECT} compute backend-services describe --global ${BACKEND}
```

* If the timeout isn't set to 1 hour then you could update it using gcloud.

## For a proper fix I think we can try something like

1. Remove checkBackend from setup_backend.sh
    * setup_backend.sh is configuring the JWT policy based on IAP; its not updating the backend timeout so there's no reason to check that

   * i.e delete everything after https://github.com/kubeflow/manifests/blob/56e2fb15db286198f7a53723cb1fbfecf3fe83fb/gcp/iap-ingress/base/config-map.yaml#L113
   * Add a sleep at the end for 1800-3600 seconds

I think by design we want the container to periodically restart. I think this is because we saw weird issues with IAM where we needed to restart the container to pick up IAM changes.

**Comment by jlewi** [comment]
@mattnworb any interest in submitting a PR to try improve the backend scripts; e.g.

* Add `set e` so we get better logging
* Remove the backend timeout check from setup_backend.sh

**Comment by holdenk** [comment]
I'm running into this on 0.7 RC1.

> holden@miniholden:~/repos/sparklingml$ NAMESPACE=istio-system
> holden@miniholden:~/repos/sparklingml$ SERVICE=istio-ingressgateway
> holden@miniholden:~/repos/sparklingml$ 
> holden@miniholden:~/repos/sparklingml$ kubectl -n ${NAMESPACE} get policy ingress-jwt -o yaml
> apiVersion: authentication.istio.io/v1alpha1
> kind: Policy
> metadata:
>   annotations:
>     kubectl.kubernetes.io/last-applied-configuration: |
>       {"apiVersion":"authentication.istio.io/v1alpha1","kind":"Policy","metadata":{"annotations":{},"labels":{"app.kubernetes.io/component":"iap-ingress","app.kubernetes.io/instance":"iap-ingress-v0.7.0","app.kubernetes.io/managed-by":"kfctl","app.kubernetes.io/name":"iap-ingress","app.kubernetes.io/part-of":"kubeflow","app.kubernetes.io/version":"v0.7.0","kustomize.component":"iap-ingress"},"name":"ingress-jwt","namespace":"istio-system"},"spec":{"origins":[{"jwt":{"audiences":["TO_BE_PATCHED"],"issuer":"https://cloud.google.com/iap","jwksUri":"https://www.gstatic.com/iap/verify/public_key-jwk","jwtHeaders":["x-goog-iap-jwt-assertion"],"trigger_rules":[{"excluded_paths":[{"exact":"/healthz/ready"},{"prefix":"/.well-known/acme-challenge"}]}]}}],"principalBinding":"USE_ORIGIN","targets":[{"name":"istio-ingressgateway","ports":[{"number":80}]}]}}
>   creationTimestamp: "2019-10-15T12:14:18Z"
>   generation: 2
>   labels:
>     app.kubernetes.io/component: iap-ingress
>     app.kubernetes.io/instance: iap-ingress-v0.7.0
>     app.kubernetes.io/managed-by: kfctl
>     app.kubernetes.io/name: iap-ingress
>     app.kubernetes.io/part-of: kubeflow
>     app.kubernetes.io/version: v0.7.0
>     kustomize.component: iap-ingress
>   name: ingress-jwt
>   namespace: istio-system
>   resourceVersion: "5925"
>   selfLink: /apis/authentication.istio.io/v1alpha1/namespaces/istio-system/policies/ingress-jwt
>   uid: 4fc12317-ef45-11e9-a1ec-42010a8400b2
> spec:
>   origins:
>   - jwt:
>       audiences:
>       - /projects/658583755711/global/backendServices/4483781589608171228
>       issuer: https://cloud.google.com/iap
>       jwksUri: https://www.gstatic.com/iap/verify/public_key-jwk
>       jwtHeaders:
>       - x-goog-iap-jwt-assertion
>       trigger_rules:
>       - excluded_paths:
>         - exact: /healthz/ready
>         - prefix: /.well-known/acme-challenge
>   principalBinding: USE_ORIGIN
>   targets:
>   - name: istio-ingressgateway
>     ports:
>     - number: 80
> 

Describing the baclend service:

> 
> holden@miniholden:~/repos/sparklingml$ gcloud  compute backend-services describe --global 4483781589608171228
> affinityCookieTtlSec: 0
> backends:
> - balancingMode: RATE
>   capacityScaler: 1.0
>   group: https://www.googleapis.com/compute/v1/projects/boos-demo-projects-are-rad/zones/europe-west1-b/instanceGroups/k8s-ig--d06d8cfe13f2d356
>   maxRatePerInstance: 1.0
> connectionDraining:
>   drainingTimeoutSec: 0
> creationTimestamp: '2019-10-15T05:14:43.775-07:00'
> description: '{"kubernetes.io/service-name":"istio-system/istio-ingressgateway","kubernetes.io/service-port":"80"}'
> enableCDN: false
> fingerprint: Ae1G-avBKI4=
> healthChecks:
> - https://www.googleapis.com/compute/v1/projects/boos-demo-projects-are-rad/global/healthChecks/k8s-be-31380--d06d8cfe13f2d356
> iap:
>   enabled: true
>   oauth2ClientId: CLIENTID
>   oauth2ClientSecretSha256: SECRETSHA256
> id: '4483781589608171228'
> kind: compute#backendService
> loadBalancingScheme: EXTERNAL
> name: k8s-be-31380--d06d8cfe13f2d356
> port: 31380
> portName: port31380
> protocol: HTTP
> selfLink: https://www.googleapis.com/compute/v1/projects/boos-demo-projects-are-rad/global/backendServices/k8s-be-31380--d06d8cfe13f2d356
> sessionAffinity: NONE
> timeoutSec: 30
> 

I tried bumping the timeout as suggested by @jlewi but its still failing.

**Comment by jlewi** [comment]
@holdenk The fix https://github.com/kubeflow/manifests#513 didn't go in until after you filed your bug report.

Also note the fix is in kubeflow/manifests so its not just about the kfctl version you are using but the version of kubeflow/manifests you pick up.

**Comment by holdenk** [comment]
So I just redployed. Looking in my manifests now I can see https://github.com/kubeflow/manifests/pull/531/files is included.

The error message is different but the IAP enabler still isn't working:

> backend name is 
> + sleep 2
> + [[ -z '' ]]
> ++ kubectl --namespace=istio-system get ingress envoy-ingress -o 'jsonpath={.metadata.annotations.ingress\.kubernetes\.io/backends}'
> + BACKENDS=
> + echo 'fetching backends info with envoy-ingress: '
> fetching backends info with envoy-ingress: 
> ++ grep -o 'k8s-be-31380--[0-9a-z]\+'
> ++ echo
> + BACKEND_NAME=
> + echo 'backend name is '
> + sleep 2
> backend name is 
> + [[ -z '' ]]
> ++ kubectl --namespace=istio-system get ingress envoy-ingress -o 'jsonpath={.metadata.annotations.ingress\.kubernetes\.io/backends}'
> + BACKENDS=
> + echo 'fetching backends info with envoy-ingress: '
> fetching backends info with envoy-ingress: 
> ++ echo
> ++ grep -o 'k8s-be-31380--[0-9a-z]\+'
> + BACKEND_NAME=
> + echo 'backend name is '
> + sleep 2
> backend name is 
> + [[ -z '' ]]
> ++ kubectl --namespace=istio-system get ingress envoy-ingress -o 'jsonpath={.metadata.annotations.ingress\.kubernetes\.io/backends}'
> + BACKENDS=
> + echo 'fetching backends info with envoy-ingress: '
> fetching backends info with envoy-ingress: 
> ++ grep -o 'k8s-be-31380--[0-9a-z]\+'
> ++ echo
> + BACKEND_NAME=
> + echo 'backend name is '
> + sleep 2
> backend name is 
> + [[ -z '' ]]
> ++ kubectl --namespace=istio-system get ingress envoy-ingress -o 'jsonpath={.metadata.annotations.ingress\.kubernetes\.io/backends}'
> + BACKENDS=
> fetching backends info with envoy-ingress: 
> + echo 'fetching backends info with envoy-ingress: '
> ++ ++ echo
> grep -o 'k8s-be-31380--[0-9a-z]\+'
> + BACKEND_NAME=
> + echo 'backend name is '
> backend name is 
> + sleep 2
> + [[ -z '' ]]
> ++ kubectl --namespace=istio-system get ingress envoy-ingress -o 'jsonpath={.metadata.annotations.ingress\.kubernetes\.io/backends}'
> + BACKENDS=
> fetching backends info with envoy-ingress: 
> + echo 'fetching backends info with envoy-ingress: '
> ++ grep -o 'k8s-be-31380--[0-9a-z]\+'
> ++ echo
> + BACKEND_NAME=
> + echo 'backend name is '
> + sleep 2
> backend name is 
> + [[ -z '' ]]
> ++ kubectl --namespace=istio-system get ingress envoy-ingress -o 'jsonpath={.metadata.annotations.ingress\.kubernetes\.io/backends}'
> + BACKENDS=
> + echo 'fetching backends info with envoy-ingress: '
> fetching backends info with envoy-ingress: 
> ++ grep -o 'k8s-be-31380--[0-9a-z]\+'
> ++ echo
> backend name is 
> + BACKEND_NAME=
> + echo 'backend name is '
> + sleep 2
> + [[ -z '' ]]
> ++ kubectl --namespace=istio-system get ingress envoy-ingress -o 'jsonpath={.metadata.annotations.ingress\.kubernetes\.io/backends}'
> + BACKENDS=
> + echo 'fetching backends info with envoy-ingress: '
> fetching backends info with envoy-ingress: 
> ++ grep -o 'k8s-be-31380--[0-9a-z]\+'
> ++ echo
> + BACKEND_NAME=
> + echo 'backend name is '
> + sleep 2
> backend name is 
> + [[ -z '' ]]
> ++ kubectl --namespace=istio-system get ingress envoy-ingress -o 'jsonpath={.metadata.annotations.ingress\.kubernetes\.io/backends}'
> + BACKENDS=
> + echo 'fetching backends info with envoy-ingress: '
> fetching backends info with envoy-ingress: 
> ++ grep -o 'k8s-be-31380--[0-9a-z]\+'
> ++ echo
> + BACKEND_NAME=
> + echo 'backend name is '
> + sleep 2
> backend name is 
> + [[ -z '' ]]
> ++ kubectl --namespace=istio-system get ingress envoy-ingress -o 'jsonpath={.metadata.annotations.ingress\.kubernetes\.io/backends}'
> fetching backends info with envoy-ingress: 
> + BACKENDS=
> + echo 'fetching backends info with envoy-ingress: '
> ++ echo++ grep -o 'k8s-be-31380--[0-9a-z]\+'
> 
> backend name is 
> + BACKEND_NAME=
> + echo 'backend name is '
> + sleep 2
> + [[ -z '' ]]
> ++ kubectl --namespace=istio-system get ingress envoy-ingress -o 'jsonpath={.metadata.annotations.ingress\.kubernetes\.io/backends}'
> + BACKENDS=
> + echo 'fetching backends info with envoy-ingress: '
> fetching backends info with envoy-ingress: 
> ++ ++ grep echo
> -o 'k8s-be-31380--[0-9a-z]\+'
> backend name is 
> + BACKEND_NAME=
> + echo 'backend name is '
> + sleep 2
> + [[ -z '' ]]
> ++ kubectl --namespace=istio-system get ingress envoy-ingress -o 'jsonpath={.metadata.annotations.ingress\.kubernetes\.io/backends}'
> + BACKENDS=
> + echo 'fetching backends info with envoy-ingress: '
> fetching backends info with envoy-ingress: 
> ++ grep -o 'k8s-be-31380--[0-9a-z]\+'
> ++ echo
> + BACKEND_NAME=
> + echo 'backend name is '
> + sleep 2
> backend name is 
> + [[ -z '' ]]
> ++ kubectl --namespace=istio-system get ingress envoy-ingress -o 'jsonpath={.metadata.annotations.ingress\.kubernetes\.io/backends}'
> fetching backends info with envoy-ingress: 
> + BACKENDS=
> + echo 'fetching backends info with envoy-ingress: '
> ++ grep -o 'k8s-be-31380--[0-9a-z]\+'
> ++ echo
> + BACKEND_NAME=
> + echo 'backend name is '
> + sleep 2
> backend name is 
> + [[ -z '' ]]
> ++ kubectl --namespace=istio-system get ingress envoy-ingress -o 'jsonpath={.metadata.annotations.ingress\.kubernetes\.io/backends}'
> + BACKENDS=
> + echo 'fetching backends info with envoy-ingress: '
> fetching backends info with envoy-ingress: 
> ++ grep -o 'k8s-be-31380--[0-9a-z]\+'
> ++ echo
> + BACKEND_NAME=
> backend name is 
> + echo 'backend name is '
> + sleep 2
>

**Comment by jlewi** [comment]
@holdenk Your problem looks different. Can you provide the output of 

```
kubectl -n istio-system ingress envoy-ingress -o yaml
kubectl -n istio-system describe envoy-ingress
```

It looks to me like the BACKEND_NAME isn't getting set

```
++ kubectl --namespace=istio-system get ingress envoy-ingress -o 'jsonpath={.metadata.annotations.ingress.kubernetes.io/backends}'
BACKENDS=
```

So I suspect your ingress is failing to create the GCP backend for some other reason (insufficient quota)?

It would be good to confirm if that is the problem. If it is we should open up a separate issue to provide a more helpful error message in the logs.

**Comment by jlewi** [comment]
Fixed in 0.7.0
